### PR TITLE
ast&cmd: Only enforcing `schemas` annotations if `--schema` flag is used

### PR DESCRIPTION
**Pull request by johanfylling**
Impacts the `eval` and `check` commands.

Fixes: #5430
Signed-off-by: Johan Fylling <johan.dev@fylling.se>

### CODE REVIEW

**Review by anderseknert — DISMISSED** [review]
Could there be other commands that might trigger this condition?
Like `fmt`, `inspect`, `build`, `deps`, `exec`, `run`, `parse`, and `bench` 😅 I'd imagine at least a few of them runs the compiler, but I'm not sure whether they process annotations in a way that would have this error triggered. Do you know?

If that's confirmed not to be the case, LGTM 👍

File: cmd/eval_test.go [hunk 1]
```diff
@@ -366,6 +366,67 @@ func testEvalWithInvalidSchemaFile(input string, query string, schema string) er
...
+		if !defined || err != nil {
+			err = fmt.Errorf("Unexpected error or undefined from evaluation: %v", err)
+		}
+	})
+
+	return err
+}
+
+// Assert that 'schemas' annotations are only informing the type checker when the --schema flag is used
+func TestEvalWithSchemasAnnotationButNoSchemaFlag(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (johanfylling) [reply cmd/eval_test.go:398]: we have https://github.com/open-policy-agent/opa/blob/9bd1bfb4e1440009be1e8d1a126d35d091d7f043/cmd/eval_test.go#L308, but it looks like that one doesn't use the `schemas` annotation ..
I'll add a test.

**Comment by johanfylling** [comment]
With this change, `eval` and `check` will use annotations when type-checking; all other commands will ignore the `schemas` annotation.

**Comment by johanfylling** [comment]
Also, if you provide a `ShemaSet` in the options to `rego.New()` in the SDK, you'll enable application of the `schemas` annotation.

File: ast/compile.go [hunk 2]
```diff
@@ -407,6 +408,12 @@ func (c *Compiler) WithKeepModules(y bool) *Compiler {
 	return c
 }
 
+// WithUseTypeCheckAnnotations use schema annotations during type checking
+func (c *Compiler) WithUseTypeCheckAnnotations(useTypeCheckAnnotations bool) *Compiler {
+	c.useTypeCheckAnnotations = useTypeCheckAnnotations
```
Comment (srenatus) [inline ast/compile.go:413]:
```suggestion
func (c *Compiler) WithUseTypeCheckAnnotations(enabled bool) *Compiler {
	c.useTypeCheckAnnotations = enabled
```

File: cmd/eval_test.go [hunk 3]
```diff
@@ -305,24 +305,32 @@ func TestEvalWithInvalidInputFile(t *testing.T) {
...
 }
 
-func testEvalWithSchemaFile(t *testing.T, input string, query string, schema string) error {
+func testEvalWithSchemaFile(t *testing.T, input string, query string, schema string, policy string, expTypeErr bool) error {
 	files := map[string]string{
 		"input.json":  input,
 		"schema.json": schema,
 	}
 
+	policyFilePresent := len(policy) > 0
```
Comment (srenatus) [inline cmd/eval_test.go:314]: [nit/style] I'd use this to give an indication that we're dealing with a string here, not some slice. But it's not important:
```suggestion
	policyFilePresent := policy != ""
```

File: cmd/eval_test.go [hunk 4]
```diff
@@ -332,6 +340,15 @@ func testEvalWithSchemaFile(t *testing.T, input string, query string, schema str
 			t.Fatal(err)
 		}
 
+		if expTypeErr {
+			if len(output.Errors) != 1 || output.Errors[0].Code != "rego_type_error" {
+				err = fmt.Errorf("expected type conflict")
```
Comment (srenatus) [inline cmd/eval_test.go:345]:
```suggestion
				err = fmt.Errorf("expected type conflict, got %v", output.Errors)
```

File: cmd/eval_test.go [hunk 5]
```diff
@@ -332,6 +340,15 @@ func testEvalWithSchemaFile(t *testing.T, input string, query string, schema str
...
+			} else {
+				err = nil
```
Comment (srenatus) [inline cmd/eval_test.go:347]: [nit] I don't think we need this
```suggestion
```

**Review by srenatus — APPROVED** [review]
LGTM nitpicks only

File: cmd/eval_test.go (hunk 5 above)
Comment (johanfylling) [reply cmd/eval_test.go:347]: Actually, `err` could contain an expected error from the call to `eval`. But I think readability is improved if I remove this and don't reuse `err` in the above call.

**Review by srenatus — APPROVED** [review]
Thanks

**Review by srenatus — APPROVED** [review]
LGTM!

@johanfylling can you please summarise the change in the commit message when doing the squash&merge?

**Pull request merged**